### ISSUE TITLE
Issue 18h: Custom image classification model architecture

### ISSUE THREAD

**Issue by bmander** [bmander/thecityandthebike#68]
## Parent Issue
Part of #18 - Create image classification system

## Branch
`issue-18-h-model-architecture`

## Dependencies
- None (can start immediately, in parallel with other sub-tasks)

## Blocks
- Issue 18c (#63, training service) — the training service wraps this model in a Vertex AI Custom Training Job

## Scope
Medium

## Description

Design and implement the custom image classification model using TensorFlow with transfer learning, replacing the AutoML approach. This produces a training script that Vertex AI Custom Training will execute.

### Why Custom Instead of AutoML

- Full control over model architecture, size, and inference latency
- Can tune hyperparameters for our specific 4-class problem
- Potentially lower training cost (custom training is cheaper per node-hour than AutoML)
- Can export to TFLite or SavedModel for flexible deployment
- Reproducible training with version-controlled code

### Model Architecture

**Base: TensorFlow + transfer learning from a pre-trained ImageNet model**

Candidates to evaluate:
- **EfficientNet-B0** — good accuracy/size tradeoff, strong with limited data
- **MobileNetV2** — very lightweight, fast inference, good for small class counts
- **ResNet50** — well-understood baseline, robust transfer learning

Strategy:
1. Load pre-trained model (ImageNet weights), remove classification head
2. Add: GlobalAveragePooling2D → Dropout → Dense(4, softmax)
3. Freeze base layers initially, train head
4. Optionally unfreeze top N layers for fine-tuning
5. Output: predicted class + confidence scores for all 4 classes

### Training Script

Create a standalone training script that:
1. Accepts hyperparameters as command-line args (learning rate, epochs, batch size, base model)
2. Loads training data from a GCS path (CSV manifest pointing to images + labels)
3. Applies data augmentation (random flip, rotation, brightness, contrast)
4. Splits into train/validation sets
5. Trains with early stopping on validation loss
6. Exports a TensorFlow SavedModel to a GCS output path
7. Logs metrics (accuracy, loss, per-class precision/recall) for tracking

```
api/
├── ml/
│   ├── train.py              # Training script entry point
│   ├── model.py              # Model architecture definition
│   ├── data.py               # Data loading, augmentation, CSV parsing
│   ├── requirements.txt      # ML-specific deps (tensorflow, etc.)
│   └── Dockerfile            # Container for Vertex AI Custom Training
```

### Input Format

Same CSV format as defined in the plan:
```csv
training,gs://bucket/images/abc.jpg,lime_left
validation,gs://bucket/images/def.jpg,bird_right
```

### Image Preprocessing

- Resize to model input size (e.g., 224x224 for EfficientNet-B0)
- Normalize pixel values to [0, 1] or model-specific range
- Data augmentation (training only): random horizontal flip, rotation ±15°, brightness/contrast jitter

### Hyperparameters

| Parameter | Default | Description |
|-----------|---------|-------------|
| `base_model` | `EfficientNetB0` | Pre-trained base model |
| `learning_rate` | `1e-4` | Initial learning rate |
| `epochs` | `50` | Max training epochs |
| `batch_size` | `32` | Training batch size |
| `dropout` | `0.3` | Dropout rate before classification head |
| `fine_tune_layers` | `0` | Number of base model layers to unfreeze (0 = head only) |
| `image_size` | `224` | Input image dimension |

### Output

- TensorFlow SavedModel at `gs://bucket/ml/models/{run_id}/`
- Training metrics JSON at `gs://bucket/ml/models/{run_id}/metrics.json`
- The SavedModel accepts base64-encoded image bytes and returns class probabilities

### Serving Signature

The model must have a serving signature compatible with Vertex AI Prediction:
- Input: `{"instances": [{"content": "<base64-image>"}]}`
- Output: `{"predictions": [{"displayNames": [...], "confidences": [...]}]}`

This may require a custom serving container or a `@tf.function` serving wrapper.

### Vertex AI Custom Training Integration

The training script runs as a Vertex AI Custom Training Job:
```python
job = aiplatform.CustomTrainingJob(
    display_name=f"tcatb-fender-training-{run_id}",
    script_path="ml/train.py",
    container_uri="us-docker.pkg.dev/vertex-ai/training/tf-gpu.2-14.py310:latest",
    requirements=["tensorflow-io"],  # any extra deps
    model_serving_container_image_uri="us-docker.pkg.dev/vertex-ai/prediction/tf2-gpu.2-14:latest",
)

model = job.run(
    dataset=dataset,
    model_display_name=f"tcatb-fender-model-{run_id}",
    args=["--epochs", "50", "--learning_rate", "1e-4", "--base_model", "EfficientNetB0"],
    replica_count=1,
    machine_type="n1-standard-4",
    accelerator_type="NVIDIA_TESLA_T4",
    accelerator_count=1,
)
```

### Local Development & Testing

- Training script should work locally with `ML_BACKEND=mock` or a small local dataset
- Unit tests for model architecture (correct input/output shapes)
- Unit tests for data loading and augmentation
- Small integration test: train on 5 images per class for 2 epochs, verify SavedModel exports

### Files

**New:**
- `api/ml/train.py` — training entry point
- `api/ml/model.py` — model architecture
- `api/ml/data.py` — data loading and augmentation
- `api/ml/requirements.txt` — ML dependencies (tensorflow, etc.)
- `api/ml/Dockerfile` — custom training container (if needed beyond pre-built)

### Tests

- Model architecture tests (input shape → output shape with 4 classes)
- Data loading tests (CSV parsing, image loading, augmentation)
- Small-scale training test (2 epochs on tiny dataset, verify convergence direction)
- SavedModel export and reload test
- Serving signature compatibility test
